witness/golang/cmd/witness/internal/witness/witness_test.go

```diff
@@ -0,0 +1,287 @@
+	t.Helper()
+	db, err := sql.Open("sqlite3", ":memory:")
+	if err != nil {
+		t.Fatalf("failed to open temporary in-memory DB: %v", err)
+	}
+	return db, db.Close
+}
+
+func TestGetChkpt(t *testing.T) {
+	for _, test := range []struct {
+		desc    string
+		setID   string
+		queryID string
+		c       Chkpt
+		want    string
+	}{
+		{
+			desc:    "happy path",
+			setID:   "testlog",
+			queryID: "testlog",
+			c:       initChkpt,
+			want:    "there",
+		}, {
+			desc:    "other log",
+			setID:   "testlog",
+			queryID: "otherlog",
+			c:       initChkpt,
+			want:    "not there",
+		}, {
+			desc:    "nothing there",
+			setID:   "testlog",
+			queryID: "testlog",
+			c:       initChkpt,
+			want:    "not there",
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			d, closeFn := mustCreateDB(t)
+			defer closeFn()
+			ctx := context.Background()
+			// Set up log keys and sign checkpoint.
+			logSK, logPK, err := note.GenerateKey(rand.Reader, test.setID)
+			if err != nil {
+				t.Errorf("couldn't generate log keys: %v", err)
+			}
+			signed, err := signChkpts(logSK, []string{string(test.c.Raw)})
+			if err != nil {
+				t.Errorf("couldn't sign checkpoint: %v", err)
+			}
+			test.c.Raw = signed[0]
+			// Set up witness keys and other parameters.
+			wSK, wPK, err := note.GenerateKey(rand.Reader, "witness")
+			if err != nil {
+				t.Errorf("couldn't generate witness keys: %v", err)
+			}
+			opts, err := newOpts(d, test.setID, logPK, wSK)
+			if err != nil {
+				t.Errorf("couldn't create witness opts: %v", err)
+			}
+			w, err := New(opts)
+			if err != nil {
+				t.Fatalf("couldn't create witness: %v", err)
+			}
+			// Set a checkpoint for the log if we want to for this test.
+			if test.desc != "nothing there" {
+				if _, err := w.Update(ctx, test.setID, test.c.Raw, [][]byte{}); err != nil {
+					t.Errorf("failed to set checkpoint: %v", err)
+				}
+			}
+			// Try to get the latest checkpoint.
+			cosigned, err := w.GetCheckpoint(test.queryID)
+			if test.want == "not there" && err == nil {
+				t.Fatalf("returned a checkpoint but shouldn't have")
+			}
+			// If we got something then verify it under the log and
+			// witness public keys.
+			if test.want == "there" {
+				if err != nil {
+					t.Errorf("failed to get latest: %v", err)
+				}
+				wV, err := note.NewVerifier(wPK)
+				if err != nil {
+					t.Errorf("couldn't create a witness verifier: %v", err)
+				}
+				logV, err := note.NewVerifier(logPK)
+				if err != nil {
+					t.Errorf("couldn't create a log verifier: %v", err)
+				}
+				n, err := note.Open(cosigned, note.VerifierList(logV, wV))
+				if err != nil {
+					t.Errorf("couldn't verify the co-signed checkpoint: %v", err)
+				}
+				if len(n.Sigs) != 2 {
+					t.Fatalf("checkpoint doesn't verify under enough keys")
+				}
+			}
+		})
+	}
+}
+
+func TestUpdate(t *testing.T) {
+	for _, test := range []struct {
+		desc  string
+		initC Chkpt
+		newC  Chkpt
+		pf    [][]byte
+		want  string
```

Comment:
This might be better as  `bool wantErr`?

Reply:
Yup, same, done.